Update process stats

tlrx opened the pull request:
This pull request:
- removes the sigar-specific process stats: `process.mem.resident`, `process.mem.share`
- add default implementation based on JMX beans for all other process stats
- make ProcessProbe a singleton so that it can be loaded before security manager
- move some process fields from Nodes Info to Nodes Stats API (max file descriptors, mlockall)
- add documentation for process stats & process info

tlrx commented:
I'm updating the Nodes Stats and Nodes Info API piece by piece... Here is the `process` part. 

@kimchy @rmuir @clintongormley Can you have a look please? thanks.

kimchy commented on core/src/main/java/org/elasticsearch/monitor/process/ProcessProbe.java:
```diff
@@ -19,12 +19,249 @@
+        } catch (Exception e) {
+            // not available
+        }
+        getMaxFileDescriptorCountField = method;
```
similar comment from my OS level review, extract it to a method to make sure we don't leak the method down the code

tlrx replied:
Done.

kimchy commented:
left a couple of notes, more thoughts:

- I think we should keep the mlockall flag in ProcessInfo.
- Only have getters in Info/Stats
- I suggest we skip getting the user/sys cpu time for the process using threads, I am not terribly sure about it, we can alway add it afterwards?
-

rmuir commented on core/src/main/java/org/elasticsearch/bootstrap/Bootstrap.java:
```diff
@@ -143,14 +143,22 @@ public boolean handle(int code) {
         StringHelper.randomId();
     }
 
+    public static void initializeProbes() {
```
Can this be package private? I think its not needed except by bootstrap and bootstrap for testing.

tlrx replied:
Ok

rmuir commented on core/src/test/java/org/elasticsearch/monitor/process/ProcessProbeTests.java:
```diff
@@ -0,0 +1,57 @@
+    public void testProcessInfo() {
+        ProcessInfo info = probe.processInfo();
+        assertNotNull(info);
+        assertThat(info.getRefreshInterval(), anyOf(equalTo(-1L), greaterThanOrEqualTo(0L)));
```
maybe in a followup we can think about removing these -1s... see what platforms fail, and better fine-grain the stuff (e.g. add assumption for WINDOWS, IBM jdk, whatever it might be). Then we know when and where stats are available.

tlrx commented:
@kimchy @rmuir I updated the code with your comments. Can you please look again? thanks.

> I suggest we skip getting the user/sys cpu time for the process using threads, I am not terribly sure about it, we can alway add it afterwards?

I removed process cpu user/sys time. I also added a `ProcessProbeBenchmark` class which produces the following results for 100 000 iterations:
```
testing 'getOpenFileDescriptorCount' method...
total [1347] ms, avg [0.01347] ms

testing 'getMaxFileDescriptorCount' method...
total [12] ms, avg [1.2E-4] ms

testing 'getTotalVirtualMemorySize' method...
total [890] ms, avg [0.0089] ms

testing 'getProcessCpuPercent' method...
total [3955] ms, avg [0.03955] ms

testing 'getProcessCpuTotalTime' method...
total [44] ms, avg [4.4E-4] ms

calculating process CPU user time with 'getAllThreadIds + getThreadUserTime' methods...
execution time [total: 2880 ms, avg: 0.0288 ms] for 100000 iterations with average result of 2.5158615E9

calculating process CPU user time with 'getAllThreadIds + getThreadUserTime(long[])' methods...
execution time [total: 2763 ms, avg: 0.02763 ms] for 100000 iterations with average result of 3.4250589E9
```

Calculating user cpu time for the process using threads takes ~0.03 ms. The last benchmark test uses `com.sun.management.ThreadMXBean` to retrieve thread user times for all threads at once.

kimchy commented on core/src/main/java/org/elasticsearch/monitor/process/ProcessProbe.java:
```diff
@@ -19,12 +19,156 @@
+            Method method = osMxBean.getClass().getDeclaredMethod(methodName);
+            method.setAccessible(true);
+            return method;
+        } catch (Exception e) {
```
can we catch Throwable here and in the rest of the code in this class?

kimchy commented:
left a minor comment, LGTM otherwise

tlrx commented:
@kimchy thanks for the review! I merged with Throwable catch and updated documentation.

@rmuir you were interested in the time the probe takes on Linux: I added the `ProcessProbeBenchmark` class, feel free to comment here if needed.

rmuir commented on core/src/test/java/org/elasticsearch/benchmark/monitor/process/ProcessProbeBenchmark.java:
```diff
@@ -0,0 +1,130 @@
+                elapsed, (elapsed / (double)ITERATIONS), ITERATIONS, (sum / (double)ITERATIONS));
+
+        /* Commented as com.sun.management is listed as forbidden usage
+        if (threadMxBean instanceof com.sun.management.ThreadMXBean) {
```
We can use the SuppressForbidden annotation on top of the class to fix this.

rmuir commented:
Thanks for investigating the performance! I think the benchmark is useful because in the past, there has been issues with stats apis being on the slow side.

kimchy commented:
++, it helps us decide if we want to add it down the road
